grails-app/domain/org/pih/warehouse/picklist/PicklistItem.groovy

```diff
@@ -96,6 +100,10 @@ class PicklistItem implements Serializable {
         return (inventoryItem ? inventoryItem.pickable : true) && (binLocation ? binLocation.pickable : true)
     }
 
+    Boolean isAutoPicked() {
+        return pickType == null || pickType == PickType.AUTO
```

Comment:
while I don't mind it, I would probably prefer to just have a migration for all existing pick list items, to have the AUTO.
Why I think it would be better is that `null` should be "reserved", so that if any of picklist items have null, we would know, that something must've gone wrong.
I don't which approach is preferred in this case, so while I don't mind your solution, I would like to know your and others opinion.

Reply:
I agree with @kchelstowski, I also think we should go with migration.

Reply:
If we do a migration we should have a plan for it. We'd want to do a select on the db to see how many production rows will be affected and also make sure that null == PickType.AUTO in 100% of cases. We don't want to break any existing data

Reply:
I agree with the database migration approach. But we also might have scenarios that we do not know what is automatic vs manual. It's possible that everything can be derived using the reason code, comment and/or last updated. But it might be also be really difficult to determine.

Reply:
> [!IMPORTANT] 
> TL;DR: I think the conclusion I've come to is that we the only way to distinguish between an "original" (automatic) vs "edited" (manual) pick is to look at whether reason code is empty. This does not actually answer the question in an authoratative way because we don't store the "original" picklist items anywhere. And it seems they are re-computed when we are rendering the comparison in the UI. 
>
> I could very well be wrong about that last point, so I need someone to correct me. But if that assertion holds, then this is not ideal because the "original" picklist items can change over time due to the nature of the FEFO algorithm and stock being updated. Therefore, knowing whether something was an original (auto allocated) picklist item becomes very difficult. 


# My Analysis
One thing that hasn't been mentioned is that there's aleady a mix of auto vs manual picklist items in the system. As far as I recall not all picklist items were automatically allocated. Some were automatically allocated and some were edits to the automatically allocated picklist items. 

Therefore, we need to find a way to distinguish between the two. That will probably require some discovery. 

> [!NOTE]
> I'm only looking at allocations based off requisitions. transfers and returns are being ignored for the moment.

```
mysql> select count(*) from (select picklist_id, requisition_item_id, count(*) from picklist_item group by picklist_id, requisition_item_id having count(*) > 1) as picklist_items;
+----------+
| count(*) |
+----------+
|    45174 |
+----------+
1 row in set (14.61 sec)
```
> [!WARNING] 
> Holy crap, 15 seconds!!  

I mentioned somewhere that the auto vs manual allocated might be derivable by a combination of properties: 
* reason code
* comment
* version?

So let's take a look at an actual example
```
mysql> select * from picklist_item where requisition_item_id = 'ff8080818715c63401874d6fa3bb46e3' \G
*************************** 1. row ***************************
                 id: ff8080818715c634018752322a585612
           quantity: 8
        picklist_id: ff8080818715c63401874d7f3e88476c
  inventory_item_id: 09f5e1167dba69cc017dbefcc5041074
            version: 0
       date_created: 2023-04-05 16:13:52
       last_updated: 2023-04-05 16:13:52
      created_by_id: NULL
      updated_by_id: NULL
             status: NULL
        reason_code: PACKAGE_SIZE
            comment: NULL
requisition_item_id: ff8080818715c63401874d6fa3bb46e3
 picklist_items_idx: NULL
    bin_location_id: ff8080816c90f81a016cbe6e87a1774d
         sort_order: 4900
      order_item_id: NULL
    quantity_picked: NULL
        date_picked: NULL
       picked_by_id: NULL
*************************** 2. row ***************************
                 id: ff8080818715c634018752322bbe5613
           quantity: 10
        picklist_id: ff8080818715c63401874d7f3e88476c
  inventory_item_id: 8a8a9e9668ca65ad0168cb77ac040946
            version: 0
       date_created: 2023-04-05 16:13:52
       last_updated: 2023-04-05 16:13:52
      created_by_id: NULL
      updated_by_id: NULL
             status: NULL
        reason_code: PACKAGE_SIZE
            comment: NULL
requisition_item_id: ff8080818715c63401874d6fa3bb46e3
 picklist_items_idx: NULL
    bin_location_id: ff8080816c90f81a016cbe6e87a1774d
         sort_order: 4900
      order_item_id: NULL
    quantity_picked: NULL
        date_picked: NULL
       picked_by_id: NULL
*************************** 3. row ***************************
                 id: ff8080818715c634018752322d2a5614
           quantity: 30
        picklist_id: ff8080818715c63401874d7f3e88476c
  inventory_item_id: 8a8a9e9666194c890166266bbd940908
            version: 0
       date_created: 2023-04-05 16:13:53
       last_updated: 2023-04-05 16:13:53
      created_by_id: NULL
      updated_by_id: NULL
             status: NULL
        reason_code: PACKAGE_SIZE
            comment: NULL
requisition_item_id: ff8080818715c63401874d6fa3bb46e3
 picklist_items_idx: NULL
    bin_location_id: ff8080816c90f81a016cbe6e87a1774d
         sort_order: 4900
      order_item_id: NULL
    quantity_picked: NULL
        date_picked: NULL
       picked_by_id: NULL
3 rows in set (0.00 sec)
```
Oof. Not what I was hoping to see but I vaguely remember this decision. 

Some observations
* There can be multiple automatically allocated picklist items for a given picklist / requisition item ID
* There can be multiple manually allocated (edited) picklist items for a given picklist / requisition item ID
* We are storing a reason code on each of the picks so it seems like we're deleting the original picklist item 
* When the UI compares the "manual" picklist items vs the "original" picklist items, we actually need to re-compute the "original" picklist items. 
* The "original" picklist items could change over time based on the FEFO algorithm.
* Having only one picklist item does not mean that the item has been auto-allocated (could be done through an import, at least that will be true in the future).
* Picklist items are either original (auto) or edited (manual), the only way to distinguish is to see if it has a reason code.